Add devcontainer

GregAC opened the pull request:
This allows a github codespaces setup for building Sonata software.

For now it points to a container that's under my user, the lowRISC org is setup such that I cannot just publish a public container image. I think if we setup some automated build and publish from this repository we can get the container under the lowrisc namespace on ghcr.io but this current setup will do the job for now.

To check out the codespace go to my branch here: https://github.com/GregAC/sonata-software/tree/devcontainer click on 'Code' and choose the 'Codespaces' tab. You should see a green 'Create codespace on devcontainer' click this and it'll setup a codespce.

It'll take a few minutes to create, you will eventually see a vscode UI. In the bottom panel (which has the terminal) you'll see a + icon in the top right of the panel click this and you should see a new terminal open that give the prompt 'Sonata /workspaces/sonata-software>' this is a nix devshell. This has the toolchain all setup so should be able to cd into '/workspace/sonata-system' then build the software following the docs: https://github.com/lowRISC/sonata-system/tree/main/sw/cheri

GregAC commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,18 @@
+  "image": "ghcr.io/gregac/sonata-software-container:latest",
+  "remoteUser": "sonata",
+  "containerUser": "sonata",
+  "onCreateCommand": "cd /workspaces && git clone https://github.com/lowRISC/sonata-system.git",
```
This is cloned here as the intent is we want to be able to send out instructions to hackathon participants ASAP so they can trial this if they want, which will mean running the cmake build commands in sonata-system.

Should we have a xmake powered build available in 'sonata-software' ASAP we can drop this bit.

GregAC commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,18 @@
+        ]
+      }
+    },
+    "terminal.integrated.defaultProfile.linux": "Nix devshell"
```
I really want the terminal you get when you open the codespace to already be a nix devshell, though haven't got this working though it is at least the default type when you create a new terminal.

hcallahan-lowrisc replied:
I found a only slightly-hack way to do this with a vscode plugin : https://github.com/lowRISC/sonata-software/compare/main...hcallahan-lowrisc:sonata-software:pr1
I don't believe the codespaces settings schema has a built in way to do this.

GregAC commented:
Oh if you give this a try I'd recommend deleting the codespace once you're done as your limited in number of free compute hours you get per month.

nbdd0121 commented:
> the lowRISC org is setup such that I cannot just publish a public container image

I think GitHub actions probably can publish container image?

nbdd0121 commented on .devcontainer/dockerfile:
The name should be `Dockerfile` per convention

GregAC commented on .devcontainer/dockerfile:
```diff
@@ -0,0 +1,45 @@
+
+# Stop nix from asking about whether we trust cache settings
+RUN mkdir -p /home/$USERNAME/.local/share/nix
+RUN echo '{"extra-substituters":{"https://nix-cache.lowrisc.org/public/":true},"extra-trusted-public-keys":{"nix-cache.lowrisc.org-public-1:O6JLD0yXzaJDPiQW1meVu32JIDViuaPtGDfjlOopU7o=":true}}' > /home/$USERNAME/.local/share/nix/trusted-settings.json
```
I put it in because when I was running `nix develop ...` it asked me to confirm various things then wrote this `trusted-settings.json` file.

You're right this is maybe surprising but I just  wanted to get something working for now.

GregAC commented:
Thanks for the review @nbdd0121 I've addressed some of your comments, the others (single user nix setup and writing the `trusted-settings.json` file) I'd prefer to leave for now if that's ok. This will work for now and it's important we get it merged sooner rather than later.
